Add Richtek RT8555 Backlight Support for samsung-gtelwifiue

person4268 opened the pull request:
The Richtek RT8555 is used as a backlight controller in samsung-gtelwifiue, and this adds support for it. It has support for both 8 bit and 10 bits of brightness levels, and this patch uses the 10 bit brightness mode.

person4268 committed:
video: backlight: Add support for Richtek RT8555 Backlight

The Richtek RT8555 is a 6 channel LED driver that supports
up to 35mA/LED. This is a backlight driver that drives ILED1.

Signed-off-by: Michael Abood <person4265@gmail.com>

TravMurav commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,151 @@
+	ret = regmap_update_bits(priv->regmap, RT8555_REG_CFG1, RT8555_EN10BIT_MASK, BIT(7));
+	if (ret)
+		return ret;
+
```
Are you sure you don't need to assert any enable pins? Would be nice to cut the power to the chip when it's unused, possibly via brightness tracking (so if it gets to 0 the power should be cut).

It's a big problem in Linux that the drivers don't take care of the power and leave things running all the time, which adds up into a much shorter battery life than could've been possible, so let's avoid that where we can

person4268 replied:
There doesn't seem to be any current limit settings, or at least anything explicitly labeled as such, there is a tuning property in the dtb, though I couldn't tell what it did. I just added some code to optionally toggle the enable pin when needed. Still WIP though, as it doesn't work, so I'm going to change this into draft for now.

TravMurav replied:
looking at the datasheet: enable pin cuts the power to the chip (as it enables/disables the internal regulator) so maybe you need to back-up/restore the state? For the limiting settings see what's available in the datasheet and look if you want to either expose any of those to the user (considering generic usage to the reasonable extent) and of course you want to fully cover any settings for your device, might want to trace the dt property reads to the register writes in the downstream driver even.

person4268 replied:
Yeah, I'll probably need to restore the state, that's likely what I was missing, since setting brightness to 0 and back did nothing, so I'll have to figure out what config I need to restore (unless it actually isn't reenabling and my code is just broken)

person4268 replied:
Fixed (6 months later)

person4268 commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,151 @@
+	ret = regmap_write(priv->regmap, RT8555_REG_ILED1_MSB, (brightness >> 8));
+	if (ret)
+		return ret;
+
```
Huh, didn't see that, probably since I don't use --strict with my checkpatch.

person4268 commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,151 @@
+	unsigned int brightness = backlight_get_brightness(bl_dev);
+	int ret;
+
+	if (brightness > RT8555_MAX_BRIGHTNESS)
```
Fixed (or it will be in a bit)

person4268 commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,151 @@
+	if (brightness > RT8555_MAX_BRIGHTNESS)
+		brightness = RT8555_MAX_BRIGHTNESS;
+
+	ret = regmap_write(priv->regmap, RT8555_REG_ILED1_LSB, (brightness & RT8555_LSB_MASK));
```
Never really heard of that, can you provide some documentation or reference? I can't seem to find anything

TravMurav replied:
Doc: https://elixir.bootlin.com/linux/latest/source/drivers/base/regmap/regmap.c#L2980

example: https://elixir.bootlin.com/linux/latest/source/drivers/iio/accel/bma400_core.c#L230 (see how the variable is used)

person4268 committed:
arm64: dts: qcom: samsung-gtelwifiue: Add RT8555 backlight driver

This adds the RT8555 backlight driver to the device tree. Using i2c-gpio
here is necessary, as the backlight ic is not on any i2c pins.

Signed-off-by: Michael Abood <person4265@gmail.com>

person4268 commented:
I've just finished rewriting this with actually proper IC initialization support. It contains commits from #241 because it'll probably get merged first and it creates a merge conflict, so I'd rather just rebase it on master once it gets merged. 

There's probably some code quality issues, but it does work properly, and it also adds the backlight to the panel in the dtb, so brightness control properly works in userspace.

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include "linux/err.h"
+#include <dt-bindings/leds/rt4831-backlight.h>
```
Please clean up unused includes.

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+	priv->dev = &client->dev;
+
+	priv->enable =
+		devm_gpiod_get_optional(&client->dev, NULL, GPIOD_OUT_HIGH);
```
Is it really worth making this GPIO optional? All the error handling required makes me think it's not.

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+		    .of_match_table = rt8555_bl_of_match },
+	.probe_new = rt8555_bl_probe,
+	.shutdown = rt8555_bl_remove
+};
```
Strange formatting, please do it similar to e.g. rt4831-backlight.c

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+	struct rt8555_priv *priv = bl_get_data(bl_dev);
+	int ret, msb, lsb;
+
+	/* If the RT8555 is disabled, there's no reason to turn it on just to read
```
Need different style for multi line comments: https://www.kernel.org/doc/html/latest/process/coding-style.html#commenting

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+
+	ret = regmap_read(priv->regmap, RT8555_REG_ILED1_LSB, &lsb);
+	if (ret)
+		return ret;
```
Consider using

```c
__le16 val;

ret = regmap_raw_read(priv->regmap, RT8555_REG_ILED1_MSB, &val, sizeof(val));
// ...

return le16_to_cpu(val);
```

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+	ret = regmap_write(priv->regmap, RT8555_REG_ILED1_MSB,
+			   (brightness >> 8));
+	if (ret)
+		return ret;
```
Similar here, consider using `regmap_raw_write` to write both at once.

stephan-gh commented on drivers/video/backlight/rt8555-backlight.c:
```diff
@@ -0,0 +1,295 @@
+{
+	struct rt8555_priv *priv = bl_get_data(bl_dev);
+	unsigned int brightness = min(backlight_get_brightness(bl_dev),
+				      priv->bl->props.max_brightness);
```
Would expect this to be checked by the backlight core already

stephan-gh commented on arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:
```diff
@@ -27,6 +27,30 @@
+		rt8555_backlight: backlight@31 {
+			status = "okay";
+			compatible = "richtek,rt8555-backlight";
+			gpio = <&msmgpio 69 GPIO_ACTIVE_HIGH>;
```
`enable-gpios` would be clearer imo

stephan-gh commented on arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:
```diff
@@ -27,6 +27,30 @@
+		#size-cells = <0>;
+
+		rt8555_backlight: backlight@31 {
+			status = "okay";
```
unneeded, is enabled by default

stephan-gh commented on arch/arm64/boot/dts/qcom/apq8016-samsung-gtelwifiue.dts:
```diff
@@ -27,6 +27,30 @@
 		};
 	};
 
+	i2c-bl {
```
i2c-backlight {

person4268 commented:
I'd also like to ask whether it'd be worth implementing a simple brightness -> perceived brightness conversion inside the driver, since currently the relation between the set brightness and the perceived brightness doesn't seem even remotely linear (the last half of the brightness values have much less of a difference in brightness than the first half). Other mainline drivers don't appear to do it, but the downstream does (though a massive lookup table). There's various functions that can estimate this (though I haven't tried), such as `pow(max_brightness, brightness/max_brightness), and the inverse of this one could be easily calculated.

stephan-gh commented:
There is something in pwm_bl for this, see the cie1931 function. I don't know if this is worth it or not or if userspace should be responsible for this.

person4268 commented:
I don't really think there'd be a clean way to implement this in userspace, personally. But then again, I'd also probably have to create the inverse of the function for the get_brightness function (or create a table like they do), which seems like a lot of effort. I'd personally rather leave such functionality out for now.

person4268 commented:
Rebased on master since #241 was merged.

stephan-gh commented:
> Is this ready for merging now?

The comment with the dt-bindings and the one potentially making the GPIO optional is still open.

Also, as per the [contribution guidelines](https://github.com/msm8916-mainline/linux/blob/master/CONTRIBUTING.md) I would appreciate if you could send this driver to the upstream mailing lists. We do not have the capacity to keep additional drivers up to date. Feel free to ask any questions about the upstreaming process. :)

stephan-gh commented:
> Personally, I'm against the making the enable GPIO non-optional, simply because other drivers leave it optional and it's not integral to the functionality of the backlight, and doesn't make the logic that much more complex, and @stephan-gh hasn't replied, so I think we can consider that closed.

My comment is not marked as "resolved" like the others and you have not replied, so it was your turn, not mine. :)

If you want to keep the GPIO optional you need to fix the error handling. The `IS_ERR` check for the GPIO must be in the probe function to detect errors. Then it can only be NULL or valid later. Functions like `gpiod_set_value()` have internal NULL checks so you can likely drop most of the if statements.

> Do I need to add myself to the MAINTAINERS file also? It only has some individual drivers, so I'm not sure.

Up to you. You can but you don't have to (not sure if it's worth it for such a small driver).

> For sending patches to the mailing list, who would I end up addressing them to, and which fields would I use? scripts/get_maintainer.pl gives me 7 email addresses. Here's the output:

I would use `--to` for the `BACKLIGHT CLASS/SUBSYSTEM` maintainers (in other words: the recipients that you would expect to apply the patch), and `--cc` everyone else, including the mailing lists. However, there is no strict rule for that, AFAIK.

> I'd assume I'd want to mail to most of the mailing lists, except for maybe fbdev, and then CC all the maintainers? Then I imagine I would have to email mainly just the devicetree mailing list for the currently nonexistent dt-bindings commit.

Please send the entire series (the two patches) to the same recipients. So you'd get the DT mailing lists and the two DT maintainers as additional Cc recipients.

person4268 commented:
> My comment is not marked as "resolved" like the others and you have not replied, so it was your turn, not mine. :)
I did reply, though?
![image](https://user-images.githubusercontent.com/28717044/178049010-9a45c45c-9d11-479b-9d1f-03c603d86820.jpeg)

It's probably the pending tag I guess?

person4268 reviewed (commented):
I think I have to leave this message so that you can see the pending comments? I didn't understand Github's review system, I guess.

I'd check some of the ones I've resolved for comments

RDS5 commented:
is anyone going to upstream this? since the backlight driver (unlike the broken panel driver) is working perfectly fine under mainline

person4268 commented:
Honestly, it's been in a perpetual state of "I want to upstream this" but I've never gotten around to actually doing it, half due to me being unsure on some properties when writing the documentation yaml (writing a description is hard when the datasheet's unclear as to what some fields mean), and half due to how completely unfamiliar I am with the LKML submission process.

I have some uncommitted WIP files I could dig out, plus I *think* an updated version of the driver that builds on newer kernels that I can put up in this PR. I also wouldn't be completely opposed to someone else continuing the PR and submitting it to the LKML in my place (though I will likely be looking into it oncemore in the coming week or so).

Also, the panel driver's broken? That's news to me. I compiled this patch against latest the pmOS kernel about 5 months ago and it was fine minus some graphical glitchiness in Plasma Mobile that wasn't there before.
